File: Bai1_DSA/Bai1.py
# Bài toán: Tìm tất cả bộ 3 số trong mảng có tổng bằng 0, không trùng nhau

from typing import List

def three_sum(nums: List[int]) -> List[List[int]]:
    # B1: Sắp xếp mảng để dễ xử lý trùng lặp và dùng hai con trỏ
    nums.sort()
    result = []
    n = len(nums)

    # B2: Duyệt từng phần tử làm mốc (nums[i])
    for i in range(n):
        # Nếu phần tử hiện tại giống phần tử trước → bỏ qua để tránh trùng
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        # Hai con trỏ left và right
        left, right = i + 1, n - 1

        # B3: Dịch chuyển hai con trỏ tìm cặp có tổng = -nums[i]
        while left < right:
            total = nums[i] + nums[left] + nums[right]

            if total == 0:
                # Ghi nhận bộ 3 thỏa điều kiện
                result.append([nums[i], nums[left], nums[right]])

                # Dịch con trỏ sang tránh trùng lặp
                left += 1
                right -= 1
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1

            elif total < 0:
                left += 1  # Tổng nhỏ hơn 0 → tăng left để tổng lớn hơn
            else:
                right -= 1  # Tổng lớn hơn 0 → giảm right để tổng nhỏ hơn

    return result


# --- Kiểm thử ---
print("Ví dụ 1:", three_sum([-1, 0, 1, 2, -1, 4]))  # [[-1, -1, 2], [-1, 0, 1]]
print("Ví dụ 2:", three_sum([0, 1, 1]))             # []
print("Ví dụ 3:", three_sum([0, 0, 0]))             # [[0, 0, 0]]

File: Bai1_DSA/Bai1.java
import java.util.*;

public class ThreeSum {
    public static List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums); // B1: Sắp xếp mảng tăng dần
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        // B2: Duyệt từng phần tử làm mốc
        for (int i = 0; i < n; i++) {
            // Bỏ qua phần tử trùng lặp
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int left = i + 1;
            int right = n - 1;

            // B3: Dùng hai con trỏ để tìm cặp còn lại
            while (left < right) {
                int total = nums[i] + nums[left] + nums[right];

                if (total == 0) {
                    // Thêm bộ 3 vào kết quả
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    left++;
                    right--;

                    // Bỏ qua các phần tử trùng ở bên trái và phải
                    while (left < right && nums[left] == nums[left - 1]) left++;
                    while (left < right && nums[right] == nums[right + 1]) right--;
                } else if (total < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return result;
    }

    // --- Hàm main để test ---
    public static void main(String[] args) {
        int[] example1 = {-1, 0, 1, 2, -1, 4};
        int[] example2 = {0, 1, 1};
        int[] example3 = {0, 0, 0};

        System.out.println("Ví dụ 1: " + threeSum(example1)); // [[-1, -1, 2], [-1, 0, 1]]
        System.out.println("Ví dụ 2: " + threeSum(example2)); // []
        System.out.println("Ví dụ 3: " + threeSum(example3)); // [[0, 0, 0]]
    }
}
